src/main/java/org/codehaus/mojo/exec/ExecMojo.java

```diff
@@ -233,6 +233,10 @@ public void execute()
             }
 
             executable = findExecutableArtifact().getFile().getAbsolutePath();
+
+            // At Unix-like systems artifacts come to the local Maven repository without "executable" bit set.
+            new File( executable ).setExecutable( true );
+
```

Comment:
I understand in theory, but in practice do not see either risk: Nothing is executed or made executable by default. It is the author of the POM who decides explicitly to execute a dependency he chooses. So it is clear that this dependency must be executable and will get executed.

Reply:
@khmarbaise, IMHO an extra parameter will complicate the plugin use without solving any practical problem. From security perspective a script or a native binary is not different from a compiled Java bytecode coming as a Maven plugin specified in POM. It is explicitly requested by the POM author as @mkarg said.

This branch of the code runs only for executableDependency when it is going to be run by "exec" goal. So it neither adds side effect to other code paths.

In the target code path it has a side effect - it permanently changes permissions to a file in the local Maven repo. But it shouldn't affect anything more:

- For the plugin it is idempotent.
- For other POMs run by the same user it shouldn't have any effects - it doesn't affect their ability to read or write the file.
- For other processes run by the same user it shouldn't add treats - this file was executed _already_, so all possible nasty things are already done.
- For other users at the same machine it has no effect, because the executable bit is set for the owner only.

Reply:
How about copying the file if it is not in the outputDirectory (aka `target/` )

Reply:
It's possible to copy the executable file to another working directory. However

- it has to be done just before each execution (`target` is ephemeral after all) impacting performance;
- it must consider concurrency (`exec:exec` is declared thread-safe) and security issues (links, random files overwrite etc);
- it implies an additional configuration parameter - the target directory for the executable file.

Not sure the gain is worth the effort and complication. I'd rather use the dependency plugin to get the file and execute it directly. While `exec:exec` with a dependency is a nice shortcut for that case.

Reply:
Maven repositories are immutable, i.e. once artifacts are deployed, they're never touched anymore. Only metadata files change. That's the reason why I was thinking about this. A temp file would work for me too.

Reply:
Well, executing a file from a Maven repository is to a little extent a hack. Maven does not retain file name (`argv[0]` may affect the program behavior), file attributes (required to execute the file). Albeit this hack is useful. The natural intent is to make it less "hackish".

Technically file system attributes is the file metadata. But fairly speaking I don't like touching files in the local repository either.

However if we plan to respect repository immutability and avoid any direct file level operations at artifacts, we'll have to do that consistently. For `exec:exec` with a dependency it means that the executable file must be copied to a work directory (a parameter) with a configurable file name (another parameter) and executed there. Regardless the executable type (a native binary or a script) and the platform. It makes even more sense because many executables consist of several files.

But such a change will be larger, more invasive and backward incompatible for certain cases. Not sure if I can make it cleanly and properly tested at relevant platforms soon.